### PR TITLE
Test checker and test case for Issue 948

### PR DESCRIPTION
**Pull request by jyluo**

### CODE REVIEW

File: framework/tests/src/tests/nontopdefault/NTDChecker.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import java.lang.annotation.Annotation;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Set;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+
+public class NTDChecker extends BaseTypeChecker {
+    /**
+     * Copied from SubtypingChecker; cannot reuse it, because SubtypingChecker is final.
```
Comment (wmdietl) [inline framework/tests/src/tests/nontopdefault/NTDChecker.java]: Why do you need this functionality? You're not suppressing any warnings.

File: framework/build.xml [hunk 2]
```diff
@@ -492,6 +492,13 @@
         </antcall>
     </target>
 
+	<target name="nontopdefault-tests" depends="jar,build-tests"
```
Comment (wmdietl) [inline framework/build.xml]: Use of tab makes indentation inconsistent.

File: framework/tests/nontopdefault/NTDTest.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+import tests.nontopdefault.qual.*;
```
Comment (wmdietl) [inline framework/tests/nontopdefault/NTDTest.java]: Tests for a particular issue should begin with the standardized
```
// Testcase for Issue abc:
// https://github.com/typetools/checker-framework/issues/abc
```

File: framework/tests/nontopdefault/NTDTest.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+/* Hierarchy:
```
Comment (wmdietl) [inline framework/tests/nontopdefault/NTDTest.java]: Would this hierarchy be better with the NTDChecker? Or as package-doc with the qualifiers?
Like this, it is unlikely to be kept in sync with the defaults specified with the qualifiers.

File: framework/tests/nontopdefault/NTDTest.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+ *    NTDTop (default for local variables, implicit upper bound, and receiver)
+ *    /     \
+ * NTDSide  NTDMiddle (default in hierarchy, default for exceptions and resource variables)
+ *   \       /
+ *   NTDBottom (default for implicit and explicit lower bounds, implicit for null literal and Void.class)
+ */
+
+// Because classes and interfaces are by default NTDMiddle, a single override is defined in the
```
Comment (wmdietl) [inline framework/tests/nontopdefault/NTDTest.java]: Why is this comment relevant here? I don't see any use of NTDMiddle.

File: framework/tests/src/tests/nontopdefault/NTDVisitor.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import org.checkerframework.javacutil.AnnotationUtils;
+import tests.nontopdefault.qual.NTDBottom;
+import tests.nontopdefault.qual.NTDMiddle;
+
+public class NTDVisitor extends BaseTypeVisitor<NTDAnnotatedTypeFactory> {
+    public NTDVisitor(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    protected final AnnotationMirror MIDDLE =
```
Comment (wmdietl) [inline framework/tests/src/tests/nontopdefault/NTDVisitor.java]: Field seems unused.

File: framework/tests/nontopdefault/NTDTest.java (hunk 5 above)
Comment (jyluo) [inline framework/tests/nontopdefault/NTDTest.java]: Everything is by default NTDMiddle, while local variables are NTDTop so `Demand<InnerDefaultReceiver> constructorReference` is actually trying to declare '@NTDTop Demand<@NTDMiddle InnerDefaultReceiver> constructorReference`. The override allows the reference to be declared.

File: framework/build.xml (hunk 2 above)
Comment (jyluo) [inline framework/build.xml]: fixed

File: framework/tests/nontopdefault/NTDTest.java (hunk 3 above)
Comment (jyluo) [inline framework/tests/nontopdefault/NTDTest.java]: fixed

File: framework/tests/nontopdefault/NTDTest.java (hunk 4 above)
Comment (jyluo) [inline framework/tests/nontopdefault/NTDTest.java]: I've moved the comment block to NTDChecker.

File: framework/tests/src/tests/nontopdefault/NTDChecker.java (hunk 1 above)
Comment (jyluo) [inline framework/tests/src/tests/nontopdefault/NTDChecker.java]: removed.

File: framework/tests/src/tests/nontopdefault/NTDVisitor.java (hunk 6 above)
Comment (jyluo) [inline framework/tests/src/tests/nontopdefault/NTDVisitor.java]: removed.

File: framework/tests/nontopdefault/NTDTest.java (hunk 5 above)
Comment (wmdietl) [inline framework/tests/nontopdefault/NTDTest.java]: That is a note about the implementation of the type system. It's not relevant to this particular test.

Comment (jyluo) [inline framework/tests/nontopdefault/NTDTest.java]: I've moved the note to the Visitor class.

**Pull request merged**